aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -535,27 +732,34 @@ void flip_kernel(TensorIterator& iter, const bool quantized) {
         });
     });
   } else {
-    // Special case: horizontal flip with vectorization and input is contiguous
-    // Context: horizontal flip leads to strides[0] < 0 and
-    // thus is_contiguous condition is not satisfied and non-vectorized code path is taken.
     auto output_strides = iter.strides(0);
     auto input_strides = iter.strides(1);
-    if (iter.ndim() > 0 && output_strides[0] < 0 && input_strides[0] == iter.element_size(1)) {
+    if (iter.ndim() > 0 && output_strides[0] == -iter.element_size(0) && input_strides[0] == iter.element_size(1)) {
+      // Special case: horizontal flip with vectorization and input is contiguous
+      // Context: horizontal flip leads to strides[0] < 0 and
+      // thus is_contiguous condition is not satisfied and non-vectorized code path is taken.
       auto iter_dtype = iter.dtype();
-      if (iter_dtype == kByte) {
-        return cpu_hflip_vec<uint8_t>(iter);
-      } else if (iter_dtype == kFloat) {
-        return cpu_hflip_vec<float>(iter);
-      } else if (iter_dtype == kInt) {
-        return cpu_hflip_vec<int32_t>(iter);
-      } else if (iter_dtype == kShort) {
-        return cpu_hflip_vec<int16_t>(iter);
-      } else if (iter_dtype == kLong) {
-        return cpu_hflip_vec<int64_t>(iter);
-      } else if (iter_dtype == kDouble) {
-        return cpu_hflip_vec<double>(iter);
-      }
-      // other dtypes are handled below with cpu_kernel_vec
+      // Ignoring half and bfloat16 as cpu_hflip_vec is slower than cpu_kernel_vec
+      if (isIntegralType(iter_dtype, true) || iter_dtype == kDouble || iter_dtype == kFloat) {
+        AT_DISPATCH_ALL_TYPES_AND(kBool,
+            iter_dtype, "hflip_cpu", [&iter] {
+              cpu_hflip_vec<scalar_t>(iter);
+        });
+        return;
+      }
+      // other dtypes (float16, bfloat16, complex) are handled by cpu_kernel_vec (see below)
+    } else if (iter.has_contiguous_first_dim()) {
+      // Special cases:
+      // a) channels last hflip on (N, C, H, W) and dtype=kByte, C in [2, 8]
+      // b) flip dim=-2 on (N, ..., M, C) and dtype=kByte, C in [2, 8]
+      auto output_strides = iter.strides(0);
+      auto iter_dtype = iter.dtype();
+      auto c = -output_strides[1];
+      if ((iter_dtype == at::kByte || iter_dtype == at::kChar) && (c >= 2 && c <= 8)) {
+        return cpu_hflip_channels_last_i8_C(iter);
+      }
```

Comment:
You're missing a stride check and the dtype check is unnecessary.
```suggestion
      auto output_strides = iter.strides(0);
      auto input_strides = iter.strides(1);
      auto c = -output_strides[1];
      if (c > 0 && c < 16 &&  // Must fit in an SSE register
          c == input_strides[1]) {
        return cpu_hflip_channels_last_C(iter);
      }
```